src/topology/metric_space/dilation.lean

```diff
@@ -0,0 +1,465 @@
+/-
+Copyright (c) 2022 Hanting Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Dilations of emetric and metric spaces
+Authors: Hanting Zhang
+-/
+import topology.metric_space.antilipschitz
+import data.fun_like.basic
+
+/-!
+# Dilations
+
+We define dilations, i.e., maps between emetric spaces that
+satisfy `edist (f x) (f y) = r * edist x y`.
+
+Defines `ratio f`, which is the ratio of some `f : dilation α β`.
+Note that `ratio f : ℝ≥0`, so we do not exclude the degenerate case of dilations
+which collapse into constant maps. Statements that do need strict dilations should
+say `f : dilation α β` and `hr : ratio f ≠ 0`.
+
+TODO: Introduce dilation equivs. Refactor the isometry API
+to match the `*_hom_class` API below.
+
+Since a lot of elementary properties don't require `eq_of_dist_eq_zero`
+we start setting up the theory for `pseudo_emetric_space` and
+we specialize to `pseudo_metric_space` and `metric_space` when needed.
+
+# Notes
+
+The type of dilations defined in this file are also referred to as
+"similiarties" or "similitudes" by other authors. The name `dilation` was choosen
+to match the Wikipedia name.
+
+# References
+
+- https://en.wikipedia.org/wiki/Dilation_(metric_space)
+- Marcel Berger, Geometry
+
+-/
+
+noncomputable theory
+
+open function set
+open_locale topological_space ennreal nnreal
+
+section defs
+
+variables (α : Type*) (β : Type*) [pseudo_emetric_space α] [pseudo_emetric_space β]
+
+/-- A dilation is a map that uniformly scales the edistance between any two points.  -/
+structure dilation :=
+(to_fun : α → β)
+(edist_eq' : ∃ r : ℝ≥0, ∀ x y : α, edist (to_fun x) (to_fun y) = r * edist x y)
```

Comment:
My inclination would be to store this as a field:

```suggestion
(ratio : ℝ≥0)
(edist_eq' : ∀ x y : α, edist (to_fun x) (to_fun y) = ratio * edist x y)
```
As far as I can tell this is equivalent if `nonempty α`

Reply:
Ah, I guess the zero metric is another place where that's not equivalent

Reply:
It just makes the `coe_mk` and `mk_coe` lemmas awkward. Not sure if that's reason enough to keep the `∃` though.

Reply:
If you include `ratio` as data, you lose the extensionality property of `fun_like` in the degenerate cases, unless you add an axiom to fix the value of `ratio` when the domain has no points with a finite nonzero distance.

Reply:
Yes indeed, that's what I was alluding to above. Adding an axiom that fixes `ratio` to `1` in those cases might be a good idea. That way, you can have one of `id.ratio = 1` or `0.ratio = 0`; whereas with the current implementation, I think you have neither.

Reply:
With the extra complication that, when there are infinite distances but no finite nonzero distances in the domain, a `ratio` of `1` means something different from a ratio of `0`, so you need to allow both in that case.

Reply:
Oh nvm, I see what you guys mean now

Reply:
The axiom would have to be like this though:
```lean
(ratio_eq' : if (∀ x y : α, edist x y = 0) then r = 0 else
             if (∀ x y : α, edist x y = 0) then r = 1 else true)
```
This is awkward, but I will see if it works out

Reply:
Have you made a typo there? The first two conditions are the same

Reply:
Also, `if p then q else true` is just a wacky way of saying `p → q`

Reply:
Actually, in the case where the space has only one element, the identity is also a constant map so it could either have ratio 1 or 0. Should we enforce `r > 0` then?

Reply:
Yeah, you're right, here's the fix
```
if (∀ x y : α, edist x y = 0) then r = 0 else (∀ x y : α, edist x y = ⊤) → r = 1
```